### ISSUE TITLE
Questions on Generation Training/Test

### ISSUE THREAD

**Issue by bjascob** [jcyk/gtos#5]
In  `generate/postprocess.py`, it looks like the golden_file is supposed to contain a json line starting with '# ::original ' but none of the preprocessing steps created this.  Did I miss something?  Any hints on where this should be generated from or an alternate method to get the `ref_stream`?

**Comment by jcyk** [comment]
hi

I think you mean `generation/postprocess.py`. Note the golden_file is supposed to be `*.txt.features` that are prodoced by [Step 4 here](https://github.com/jcyk/gtos/tree/master/generator_data).

**Comment by bjascob** [comment]
Yes, I mean `generate/postprocess.py`  Sorry about the confusion.  Specifically I'm talking about [this line](https://github.com/jcyk/gtos/blob/a13dba02a7877564648e5bd3de2bb23c4abd69b6/generator/postprocess.py#L277).  As you said, I'm expecting the golden file to be `test.txt.features`, however this file does not have any `::original` meta-data tags.  I grepped for this and none of my generated files do.

If I grep for `::original` in the code, the only place it shows up is in `postprocess.py` so the code here is definitely not generating this tag.  Looks like the reader is parsing a json string which is confusing to me with the `join` on the line below (joining all the keys?).  Can you show me the code that creates this line or let me know the format and how to reproduce it? 

Note that I'm using LDC2015E86 and don't have LDC2017T10 available.  Were these lines in the original files?  I don't see them with samples of other LDC/AMR datasets.

**Comment by jcyk** [comment]
Hi 

the `::original` lines are generated by the preprocessing steps, see [here](https://github.com/jcyk/gtos/blob/276394771c611c1cabc84c461a649259f68d32a8/generator_data/stog/data/dataset_readers/amr_parsing/amr.py#L91-L115). I am sure these lines are not in the original files.
Not sure if your problem is caused by the different versions of AMR datasets.

**Comment by jcyk** [comment]
@bjascob I think you should first make sure the preprocess steps work as expected. 
I paste a sample instance here (from `test.txt.features`)

```
# ::id nw.chtb_0325.10 ::date 2012-12-29T22:53:20 ::annotator SDL-AMR-09 ::preferred
# ::snt ( End )
# ::original ["(", "End", ")"]
# ::tokens ["(", "End", ")"]
# ::lemmas ["(", "end", ")"]
# ::pos_tags ["SYM", "NN", "SYM"]
# ::ner_tags ["O", "O", "O"]
# ::save-date Sat Dec 29, 2012 ::file nw_chtb_0325_10.txt
(e / end-01)
```
and its original (from `test.txt`)
```
# ::id nw.chtb_0325.10 ::date 2012-12-29T22:53:20 ::annotator SDL-AMR-09 ::preferred
# ::snt ( End )
# ::save-date Sat Dec 29, 2012 ::file nw_chtb_0325_10.txt
(e / end-01)
```

**Comment by bjascob** [comment]
I have everything but the `::original` line.  Backtracking the code a bit, that probably means that the [FeatureAnnotator](https://github.com/jcyk/gtos/blob/276394771c611c1cabc84c461a649259f68d32a8/generator_data/stog/data/dataset_readers/amr_parsing/preprocess/feature_annotator.py#L199) is returning `None` for the dictionary entry with `original`.  Not sure why that is.  I'm using pycorenlp v0.3.0, which is the latest.  I'm also using stanford-corenlp-full-2018-10-05, which I think is 3.9.2.

I'll have to dig into this a bit more when my computer with all the code, isn't tied up training a model.

Thanks for the help.  Now that I see where this comes from, I should be able to fix this easy enough.

BTW... how long does training the LDC2017 take and what kind of card.  Looks like on my TitanX (with revised batch sizes for 12GB memory) the smaller LDC2015 is going to take about 2 days which longer than I expected for this small amount of data.

**Comment by jcyk** [comment]
I am not sure about the training time. It may take like one day or more.
Note the default settings in the latest repo should reproduce our results.
The batch size is quite tricky. I mean for most batches, it works fine, but
there are some evil batches. There should be a better way for packing data
into batches.

**Comment by bjascob** [comment]
Good to know.  Batch size is definitely confusing but I see how it's calculated from the other post here.  I'm using 20K for no particular reason other than the original 66666 was exceeding my 12Gig.  We'll see how the results turn out.

I see the issue now.  Like all problems it's caused by a stupid mistake.  I've been working on the stog parser as well and I forgot that I have a link set to that code instead of the stog code here in your repo.  It looks like you've added 2 lines to `feature_annotator.py` to add `original` to the annotations that isn't in the stog github code.

Thanks for all your help.

**Comment by 311dada** [comment]
> I have everything but the `::original` line. Backtracking the code a bit, that probably means that the [FeatureAnnotator](https://github.com/jcyk/gtos/blob/276394771c611c1cabc84c461a649259f68d32a8/generator_data/stog/data/dataset_readers/amr_parsing/preprocess/feature_annotator.py#L199) is returning `None` for the dictionary entry with `original`. Not sure why that is. I'm using pycorenlp v0.3.0, which is the latest. I'm also using stanford-corenlp-full-2018-10-05, which I think is 3.9.2.
> 
> I'll have to dig into this a bit more when my computer with all the code, isn't tied up training a model.
> 
> Thanks for the help. Now that I see where this comes from, I should be able to fix this easy enough.
> 
> BTW... how long does training the LDC2017 take and what kind of card. Looks like on my TitanX (with revised batch sizes for 12GB memory) the smaller LDC2015 is going to take about 2 days which longer than I expected for this small amount of data.

Hi, how long did it take to train the model? I change the bash size to adapt to my GPU too, but it seems that the model is hard to converge. The loss decreases from 5 with such a slow speed. Thx for your help!

**Comment by bjascob** [comment]
Here's my [train.log](https://github.com/jcyk/gtos/files/5744083/train.log).  Looks like I left this one run for 24 hours / 700 epochs but it gets reasonable scores earlier than that.
Note that this is with LDC2015E86 data (which is smaller than LDC2017T10) and a [GTX Titan X](https://www.techpowerup.com/gpu-specs/geforce-gtx-titan-x.c2632) which is a fairly old card.

**Comment by 311dada** [comment]
Thx for your help. I am running on the AMR 2.0. It seems that the training is more slower. So, did u get the result on test dataset close to the bleu in the original paper?

**Comment by bjascob** [comment]
Probably a few points lower than in the paper (see the train.log I attached above for results on the dev set).
I see that I put a note in my code saying that the original parameters still showed very high loss after 24 hours of training.  Here's the parameters I used [train_params.txt](https://github.com/jcyk/gtos/files/5744610/train_params.txt) for the train.log above and the model that produced decent results.  The original params are behind the comments.  You'll notice that I made the net smaller in many cases so this is likely why the BLEU score is a bit lower.
Maybe you can get the original parameters to work if you play with the learning-rate schedule or other parameters but the above worked for me good enough for my purposes.

Good luck.  I haven't looked at this in a year so I don't really remember much about how things work.  I only spent a few days on it anyway, as the pretrained seq-to-seq (ie.. huggingface T5) models work much better anyway.
